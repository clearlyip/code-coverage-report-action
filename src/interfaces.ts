export interface Coverage {
  files: {[key: string]: CoverageFile}
  coverage: number
  timestamp: number
  basePath: string
}

export interface CoverageFile {
  relative: string
  absolute: string
  coverage: number
}

export interface Inputs {
  filename: string
  badge: boolean
  overallFailThreshold: number
  coverageColorRedMin: number
  coverageColorOrangeMax: number
  failOnNegativeDifference: boolean
}
