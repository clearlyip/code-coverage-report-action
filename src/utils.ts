import {promises as fs, constants as fsConstants} from 'fs'
import {XMLParser} from 'fast-xml-parser'
import * as artifact from '@actions/artifact'
import * as core from '@actions/core'
import * as github from '@actions/github'
import {Clover, parse as parseClover} from './reports/clover'
import {Cobertura, parse as parseCobertura} from './reports/cobertura'
import path from 'path'
import {Coverage, Inputs} from './interfaces'
import crypto from 'crypto'
import AdmZip from 'adm-zip'

const {access, readFile, mkdir} = fs

/**
 * Check if a file exists
 * @param {string} filename
 * @returns {Promise<boolean>}
 */
export async function checkFileExists(filename: string): Promise<boolean> {
  try {
    await access(filename, fsConstants.F_OK)
    return true
    // eslint-disable-next-line no-empty
  } catch (e) {}
  return false
}

/**
 * Parse XML
 * @param {string} filename
 * @returns {Promise<T | null>}
 */
export async function parseXML<T>(filename: string): Promise<T | null> {
  if (!(await checkFileExists(filename))) {
    return null
  }

  const contents = await readFile(filename, 'binary')

  return new XMLParser({
    ignoreAttributes: false,
    isArray: (name, jpath, isLeafNode, isAttribute) => {
      if (isAttribute) {
        return false
      }
      return inArray(jpath, [
        'coverage.project.package',
        'coverage.project.package.file'
      ])
    }
  }).parse(contents)
}

/**
 * Download Artifacts
 *
 * @param {string} name
 * @param {string} base
 * @returns {Promise<string|null>}
 */
export async function downloadArtifacts(
  name: string,
  base = 'artifacts'
): Promise<string | null> {
  const {token} = getInputs()
  const client = github.getOctokit(token)

  const {GITHUB_BASE_REF = '', GITHUB_REPOSITORY = ''} = process.env

  const [owner, repo] = GITHUB_REPOSITORY.split('/')

  for await (const runs of client.paginate.iterator(
    client.rest.actions.listWorkflowRunsForRepo,
    {
      owner,
      repo,
      branch: GITHUB_BASE_REF,
      status: 'success'
    }
  )) {
    for await (const run of runs.data) {
      if (run.name !== github.context.job) {
        continue
      }

      const artifacts = await client.rest.actions.listWorkflowRunArtifacts({
        owner,
        repo,
        run_id: run.id
      })
      if (artifacts.data.artifacts.length === 0) {
        continue
      }
      for await (const art of artifacts.data.artifacts) {
        if (art.expired) {
          continue
        }

        if (art.name !== formatArtifactName(GITHUB_BASE_REF)) {
          continue
        }

        const zip = await client.rest.actions.downloadArtifact({
          owner,
          repo,
          artifact_id: art.id,
          archive_format: 'zip'
        })

        const dir = path.join(__dirname, base)
        await mkdir(dir, {recursive: true})

        const adm = new AdmZip(Buffer.from(zip.data as string))
        adm.extractAllTo(dir, true)
        return dir
      }
    }
  }
  return null
}

/**
 * Upload Artifacts
 * @param {string[]} files
 * @param {string} name
 * @returns {Promise<artifact.UploadResponse>}
 */
export async function uploadArtifacts(
  files: string[],
  name: string
): Promise<artifact.UploadResponse> {
  const artifactClient = artifact.create()
  const artifactName = formatArtifactName(name)

  const rootDirectory = '.'
  const options = {
    continueOnError: false
  }

  return await artifactClient.uploadArtifact(
    artifactName,
    files,
    rootDirectory,
    options
  )
}

/**
 * Parse Coverage file
 * @param {string} filename
 * @returns {Promise<Coverage | null>}
 */
export async function parseCoverage(
  filename: string
): Promise<Coverage | null> {
  if (!(await checkFileExists(filename))) {
    return null
  }

  const ext = path.extname(filename)

  switch (ext) {
    case '.xml':
      {
        const xml = await parseXML<Cobertura | Clover>(filename)

        if (instanceOfCobertura(xml)) {
          return await parseCobertura(xml)
        } else if (instanceOfClover(xml)) {
          return await parseClover(xml)
        }
      }
      break
    default:
  }

  return null
}

export function createHash(data: crypto.BinaryLike): string {
  return crypto.createHash('sha256').update(data).digest('hex')
}

/**
 * Round a percentage
 * @param {number} percentage
 * @returns {number}
 */
export function roundPercentage(percentage: number): number {
  return Math.round((percentage + Number.EPSILON) * 100) / 100
}

export function escapeRegExp(string: string): string {
  return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&') // $& means the whole matched string
}

export function colorizePercentageByThreshold(
  percentage: number | null,
  threshold = 0,
  equalColor = 'grey'
): string {
  if (percentage === null) {
    return 'N/A'
  }
  if (percentage > threshold) {
    return `$\\textcolor{green}{\\text{${percentage.toString()}}}$%`
  } else if (percentage < threshold) {
    return `$\\textcolor{red}{\\text{${percentage.toString()}}}$%`
  }

  return `$\\textcolor{${equalColor}}{\\text{${percentage.toString()}}}$%`
}

/**
 * Determine a common base path
 *
 * @param {string[]} files
 * @param {string} separator
 * @returns {string}
 */
export function determineCommonBasePath(
  files: string[],
  separator = '/'
): string {
  /**
   * Given an index number, return a function that takes an array and returns the
   * element at the given index
   * @param {number} i
   * @return {function(!Array<*>): *}
   */
  const elAt = (i: number) => (a: string[]) => a[i]

  /**
   * Given an array of strings, return an array of arrays, containing the
   * strings split at the given separator
   */
  const splitStrings = files.map(i => i.split(separator))
  /**
   * Transpose an array of arrays:
   * Example:
   * [['a', 'b', 'c'], ['A', 'B', 'C'], [1, 2, 3]] ->
   * [['a', 'A', 1], ['b', 'B', 2], ['c', 'C', 3]]
   */
  const rotated = splitStrings[0].map((e, i) => splitStrings.map(elAt(i)))

  return (
    rotated
      //Checks of all the elements in the array are the same.
      .filter(arr => arr.every(e => e === arr[0]))
      .map(elAt(0))
      .join(separator)
  )
}

export function getInputs(): Inputs {
  const token = core.getInput('github_token', {required: true})
  const filename = core.getInput('filename')
  const markdownFilename = core.getInput('markdown_filename')
  const badge = core.getInput('badge') === 'true' ? true : false
  const overallFailThreshold = parseInt(core.getInput('overall_fail_threshold'))
  const coverageColorRedMin = parseInt(core.getInput('coverage_color_red_min'))
  const coverageColorOrangeMax = parseInt(
    core.getInput('coverage_color_orange_max')
  )
  const failOnNegativeDifference =
    core.getInput('fail_on_negative_difference') === 'true' ? true : false

  return {
    token,
    filename,
    badge,
    overallFailThreshold,
    coverageColorRedMin,
    coverageColorOrangeMax,
    failOnNegativeDifference,
    markdownFilename
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function instanceOfCobertura(object: any): object is Cobertura {
  return 'coverage' in object && 'packages' in object.coverage
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function instanceOfClover(object: any): object is Clover {
  return 'coverage' in object && 'project' in object.coverage
}

/**
 * Format Artifact Name
 * @param {string} name
 * @returns {string}
 */
export function formatArtifactName(name: string): string {
  return `coverage-${name}`.replace(/\//g, '-')
}

function inArray(needle: string, haystack: string[]): boolean {
  const length = haystack.length
  for (let i = 0; i < length; i++) {
    if (haystack[i] === needle) return true
  }
  return false
}
