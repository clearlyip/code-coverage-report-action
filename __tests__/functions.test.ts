import {addOverallRow, generateMarkdown} from '../src/functions'
import {
  expect,
  test,
  beforeEach,
  afterEach,
  beforeAll,
  afterAll,
  jest
} from '@jest/globals'
import fs from 'fs'
import path from 'path'
import crypto from 'crypto'
import {loadJSONFixture} from './utils'

let originalWriteFunction: (str: string) => boolean
const env: NodeJS.ProcessEnv = JSON.parse(JSON.stringify(process.env))
let testEnv: NodeJS.ProcessEnv = {}

beforeAll(async () => {
  originalWriteFunction = process.stdout.write
  process.stdout.write = jest.fn((str: string) => {
    //originalWriteFunction(str)
    return true
  }) as any

  const tempFileName = path.join(
    __dirname,
    `temp-${crypto.randomBytes(4).toString('hex')}.md`
  )
  await fs.promises.writeFile(tempFileName, '')
  process.env.GITHUB_STEP_SUMMARY = tempFileName
  process.env.INPUT_GITHUB_TOKEN = 'token'
  process.env.INPUT_FILENAME = 'filename.xml'
  process.env.INPUT_ARTIFACT_NAME = 'coverage-%name%'
})

afterAll(async () => {
  process.stdout.write = originalWriteFunction as unknown as (
    str: string
  ) => boolean
  await fs.promises.unlink(process.env.GITHUB_STEP_SUMMARY as string)
  process.env = {...env}
})

beforeEach(async () => {
  testEnv = {...process.env}
  await fs.promises.writeFile(process.env.GITHUB_STEP_SUMMARY as string, '')
})

afterEach(async () => {
  process.env = {...testEnv}
  await fs.promises.writeFile(process.env.GITHUB_STEP_SUMMARY as string, '')
})

test('add overall row without base coverage', async () => {
  const coverage = await loadJSONFixture('clover-parsed.json')
  const out = addOverallRow(coverage)

  expect(out).toStrictEqual({
    package: 'Overall Coverage',
    base_coverage: 'ðŸŸ¢ 50.51%'
  })
})

test('add overall row with base coverage', async () => {
  const coverage = await loadJSONFixture('clover-parsed.json')
  const out = addOverallRow(coverage, coverage)

  expect(out).toStrictEqual({
    package: 'Overall Coverage',
    base_coverage: 'ðŸŸ¢ 50.51%',
    new_coverage: 'ðŸŸ¢ 50.51%',
    difference: 'âšª 0%'
  })
})

test('Generate Base Clover Markdown', async () => {
  const coverage = await loadJSONFixture('clover-parsed.json')
  await generateMarkdown(coverage)
  expect(getStdoutWriteCalls()).toMatchSnapshot()
  expect(await getGithubStepSummary()).toMatchSnapshot()
})

test('Generate Base Cobertura Markdown', async () => {
  const coverage = await loadJSONFixture('cobertura-parsed.json')
  await generateMarkdown(coverage)
  expect(getStdoutWriteCalls()).toMatchSnapshot()
  expect(await getGithubStepSummary()).toMatchSnapshot()
})

test('Generate Diffed Clover Markdown', async () => {
  const coverage = await loadJSONFixture('clover-parsed.json')
  await generateMarkdown(coverage, coverage)
  expect(getStdoutWriteCalls()).toMatchSnapshot()
  expect(await getGithubStepSummary()).toMatchSnapshot()
})

test('Generate Diffed Cobertura Markdown', async () => {
  const coverage = await loadJSONFixture('cobertura-parsed.json')
  await generateMarkdown(coverage, coverage)
  expect(getStdoutWriteCalls()).toMatchSnapshot()
  expect(await getGithubStepSummary()).toMatchSnapshot()
})

test('Fail if overall coverage is below fail threshold', async () => {
  process.env.INPUT_OVERALL_COVERAGE_FAIL_THRESHOLD = '99'

  const coverage = await loadJSONFixture('clover-parsed.json')
  await generateMarkdown(coverage, coverage)
  expect(getStdoutWriteCalls()).toMatchSnapshot()
  expect(await getGithubStepSummary()).toMatchSnapshot()
})

test('Fail on negative difference', async () => {
  process.env.INPUT_FAIL_ON_NEGATIVE_DIFFERENCE = 'true'

  const coverage = await loadJSONFixture('clover-parsed.json')
  const coverageFail = JSON.parse(JSON.stringify(coverage))
  coverageFail.files[
    '7583809507a13391057c3aee722e422d50d961a87e2a3dbf05ea492dc6465c94'
  ].coverage = 69
  coverageFail.coverage = 49

  await generateMarkdown(coverageFail, coverage)
  expect(getStdoutWriteCalls()).toMatchSnapshot()
  expect(await getGithubStepSummary()).toMatchSnapshot()
})

test('Dont Fail on negative difference if negative_difference_threshold is set', async () => {
  process.env.INPUT_FAIL_ON_NEGATIVE_DIFFERENCE = 'true'
  process.env.INPUT_NEGATIVE_DIFFERENCE_THRESHOLD = '10'

  const coverage = await loadJSONFixture('clover-parsed.json')
  const coverageFail = JSON.parse(JSON.stringify(coverage))
  coverageFail.files[
    '7583809507a13391057c3aee722e422d50d961a87e2a3dbf05ea492dc6465c94'
  ].coverage = 69
  coverageFail.coverage = 49

  await generateMarkdown(coverageFail, coverage)
  expect(getStdoutWriteCalls()).toMatchSnapshot()
  expect(await getGithubStepSummary()).toMatchSnapshot()
})

test('Fail if negative_difference_threshold is set and exceeded', async () => {
  process.env.INPUT_FAIL_ON_NEGATIVE_DIFFERENCE = 'true'
  process.env.INPUT_NEGATIVE_DIFFERENCE_THRESHOLD = '1'

  const coverage = await loadJSONFixture('clover-parsed.json')
  const coverageFail = JSON.parse(JSON.stringify(coverage))
  coverageFail.files[
    '7583809507a13391057c3aee722e422d50d961a87e2a3dbf05ea492dc6465c94'
  ].coverage = 69
  coverageFail.coverage = 49

  await generateMarkdown(coverageFail, coverage)
  expect(getStdoutWriteCalls()).toMatchSnapshot()
  expect(await getGithubStepSummary()).toMatchSnapshot()
})

async function getGithubStepSummary(): Promise<string> {
  const tempFileName = process.env.GITHUB_STEP_SUMMARY as string
  return fs.promises.readFile(tempFileName, 'utf8')
}

function getStdoutWriteCalls(): string[] {
  const f = process.stdout.write as any

  return f.mock.calls.map((call: any) =>
    JSON.stringify(call[0], null, 2).replace(/^"|"$/g, '')
  )
}
