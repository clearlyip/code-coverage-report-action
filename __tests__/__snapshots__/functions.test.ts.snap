// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Dont Fail on negative difference if negative_difference_threshold is set 1`] = `
[
  "::debug::headCoverage: 49\\n",
  "::debug::baseCoverage: 50.51\\n",
  "::debug::overallDifferencePercentage: -1.51\\n",
  "::debug::negativeDifferenceThreshold: -10\\n",
  "Writing results to code-coverage-results.md\\n",
  "\\n",
  "::set-output name=file::code-coverage-results.md\\n",
  "\\n",
  "::set-output name=coverage::49\\n",
  "Writing job summary\\n",
]
`;

exports[`Dont Fail on negative difference if negative_difference_threshold is set 2`] = `
"# Code Coverage Report

| Package | Base Coverage | New Coverage | Difference |
| ------- | ------------- | ------------ | ---------- |
| main.ts | ğŸ”´ 0% | ğŸ”´ 0% | âšª 0% |
| reports/clover/index.ts | ğŸŸ¢ 100% | ğŸŸ¢ 100% | âšª 0% |
| reports/clover/parser/index.ts | ğŸŸ¢ 84.31% | ğŸŸ¢ 84.31% | âšª 0% |
| reports/clover/types/index.ts | ğŸŸ¢ 100% | ğŸŸ¢ 100% | âšª 0% |
| reports/cobertura/index.ts | ğŸŸ  66.67% | ğŸŸ  66.67% | âšª 0% |
| reports/cobertura/parser/index.ts | ğŸ”´ 13.33% | ğŸ”´ 13.33% | âšª 0% |
| reports/cobertura/types/index.ts | ğŸŸ¢ 100% | ğŸŸ¢ 100% | âšª 0% |
| utils.ts | ğŸŸ  70.94% | ğŸŸ  69% | ğŸ”´ -1.94% |
| **Overall Coverage** | **ğŸŸ¢ 50.51%** | **ğŸŸ¢ 49%** | **ğŸ”´ -1.51%** |

_Minimum allowed coverage is_ \`0%\`_, this run produced_ \`49%\`
"
`;

exports[`Fail if negative_difference_threshold is set and exceeded 1`] = `
[
  "::debug::headCoverage: 49\\n",
  "::debug::baseCoverage: 50.51\\n",
  "::debug::overallDifferencePercentage: -1.51\\n",
  "::debug::negativeDifferenceThreshold: -1\\n",
  "::error::utils.ts coverage difference was -1.94%25 which is below threshold of -1%25\\n",
  "Writing results to code-coverage-results.md\\n",
  "\\n",
  "::set-output name=file::code-coverage-results.md\\n",
  "\\n",
  "::set-output name=coverage::49\\n",
  "Writing job summary\\n",
]
`;

exports[`Fail if negative_difference_threshold is set and exceeded 2`] = `
"# Code Coverage Report

| Package | Base Coverage | New Coverage | Difference |
| ------- | ------------- | ------------ | ---------- |
| main.ts | ğŸ”´ 0% | ğŸ”´ 0% | âšª 0% |
| reports/clover/index.ts | ğŸŸ¢ 100% | ğŸŸ¢ 100% | âšª 0% |
| reports/clover/parser/index.ts | ğŸŸ¢ 84.31% | ğŸŸ¢ 84.31% | âšª 0% |
| reports/clover/types/index.ts | ğŸŸ¢ 100% | ğŸŸ¢ 100% | âšª 0% |
| reports/cobertura/index.ts | ğŸŸ  66.67% | ğŸŸ  66.67% | âšª 0% |
| reports/cobertura/parser/index.ts | ğŸ”´ 13.33% | ğŸ”´ 13.33% | âšª 0% |
| reports/cobertura/types/index.ts | ğŸŸ¢ 100% | ğŸŸ¢ 100% | âšª 0% |
| utils.ts | ğŸŸ  70.94% | ğŸŸ  69% | ğŸ”´ -1.94% |
| **Overall Coverage** | **ğŸŸ¢ 50.51%** | **ğŸŸ¢ 49%** | **ğŸ”´ -1.51%** |

_Minimum allowed coverage is_ \`0%\`_, this run produced_ \`49%\`
"
`;

exports[`Fail if overall coverage is below fail threshold 1`] = `
[
  "::debug::headCoverage: 50.51\\n",
  "::debug::baseCoverage: 50.51\\n",
  "::debug::overallDifferencePercentage: 0\\n",
  "::debug::negativeDifferenceThreshold: 0\\n",
  "::error::FAIL: Overall coverage of 50.51%25 below minimum threshold of 99%25.\\n",
  "Writing results to code-coverage-results.md\\n",
  "\\n",
  "::set-output name=file::code-coverage-results.md\\n",
  "\\n",
  "::set-output name=coverage::50.51\\n",
  "Writing job summary\\n",
]
`;

exports[`Fail if overall coverage is below fail threshold 2`] = `
"# Code Coverage Report

| Package | Base Coverage | New Coverage | Difference |
| ------- | ------------- | ------------ | ---------- |
| main.ts | ğŸ”´ 0% | ğŸ”´ 0% | âšª 0% |
| reports/clover/index.ts | ğŸŸ¢ 100% | ğŸŸ¢ 100% | âšª 0% |
| reports/clover/parser/index.ts | ğŸŸ¢ 84.31% | ğŸŸ¢ 84.31% | âšª 0% |
| reports/clover/types/index.ts | ğŸŸ¢ 100% | ğŸŸ¢ 100% | âšª 0% |
| reports/cobertura/index.ts | ğŸŸ  66.67% | ğŸŸ  66.67% | âšª 0% |
| reports/cobertura/parser/index.ts | ğŸ”´ 13.33% | ğŸ”´ 13.33% | âšª 0% |
| reports/cobertura/types/index.ts | ğŸŸ¢ 100% | ğŸŸ¢ 100% | âšª 0% |
| utils.ts | ğŸŸ  70.94% | ğŸŸ  70.94% | âšª 0% |
| **Overall Coverage** | **ğŸ”´ 50.51%** | **ğŸ”´ 50.51%** | **âšª 0%** |

_Minimum allowed coverage is_ \`99%\`_, this run produced_ \`50.51%\`
"
`;

exports[`Fail on negative difference 1`] = `
[
  "::debug::headCoverage: 49\\n",
  "::debug::baseCoverage: 50.51\\n",
  "::debug::overallDifferencePercentage: -1.51\\n",
  "::debug::negativeDifferenceThreshold: 0\\n",
  "::error::utils.ts coverage difference was -1.94%25 which is below threshold of 0%25\\n",
  "Writing results to code-coverage-results.md\\n",
  "\\n",
  "::set-output name=file::code-coverage-results.md\\n",
  "\\n",
  "::set-output name=coverage::49\\n",
  "Writing job summary\\n",
]
`;

exports[`Fail on negative difference 2`] = `
"# Code Coverage Report

| Package | Base Coverage | New Coverage | Difference |
| ------- | ------------- | ------------ | ---------- |
| main.ts | ğŸ”´ 0% | ğŸ”´ 0% | âšª 0% |
| reports/clover/index.ts | ğŸŸ¢ 100% | ğŸŸ¢ 100% | âšª 0% |
| reports/clover/parser/index.ts | ğŸŸ¢ 84.31% | ğŸŸ¢ 84.31% | âšª 0% |
| reports/clover/types/index.ts | ğŸŸ¢ 100% | ğŸŸ¢ 100% | âšª 0% |
| reports/cobertura/index.ts | ğŸŸ  66.67% | ğŸŸ  66.67% | âšª 0% |
| reports/cobertura/parser/index.ts | ğŸ”´ 13.33% | ğŸ”´ 13.33% | âšª 0% |
| reports/cobertura/types/index.ts | ğŸŸ¢ 100% | ğŸŸ¢ 100% | âšª 0% |
| utils.ts | ğŸŸ  70.94% | ğŸŸ  69% | ğŸ”´ -1.94% |
| **Overall Coverage** | **ğŸŸ¢ 50.51%** | **ğŸŸ¢ 49%** | **ğŸ”´ -1.51%** |

_Minimum allowed coverage is_ \`0%\`_, this run produced_ \`49%\`
"
`;

exports[`Generate Base Clover Markdown 1`] = `
[
  "::debug::headCoverage: 50.51\\n",
  "::debug::baseCoverage: undefined\\n",
  "::debug::overallDifferencePercentage: null\\n",
  "::debug::negativeDifferenceThreshold: 0\\n",
  "Writing results to code-coverage-results.md\\n",
  "\\n",
  "::set-output name=file::code-coverage-results.md\\n",
  "\\n",
  "::set-output name=coverage::50.51\\n",
  "Writing job summary\\n",
]
`;

exports[`Generate Base Clover Markdown 2`] = `
"# Code Coverage Report

| Package |    Coverage   |
| ------- | ------------- |
| main.ts | ğŸ”´ 0% |
| reports/clover/index.ts | ğŸŸ¢ 100% |
| reports/clover/parser/index.ts | ğŸŸ¢ 84.31% |
| reports/clover/types/index.ts | ğŸŸ¢ 100% |
| reports/cobertura/index.ts | ğŸŸ  66.67% |
| reports/cobertura/parser/index.ts | ğŸ”´ 13.33% |
| reports/cobertura/types/index.ts | ğŸŸ¢ 100% |
| utils.ts | ğŸŸ  70.94% |
| **Overall Coverage** | **ğŸŸ¢ 50.51%** |

_Minimum allowed coverage is_ \`0%\`_, this run produced_ \`50.51%\`
"
`;

exports[`Generate Base Cobertura Markdown 1`] = `
[
  "::debug::headCoverage: 49.83\\n",
  "::debug::baseCoverage: undefined\\n",
  "::debug::overallDifferencePercentage: null\\n",
  "::debug::negativeDifferenceThreshold: 0\\n",
  "Writing results to code-coverage-results.md\\n",
  "\\n",
  "::set-output name=file::code-coverage-results.md\\n",
  "\\n",
  "::set-output name=coverage::49.83\\n",
  "Writing job summary\\n",
]
`;

exports[`Generate Base Cobertura Markdown 2`] = `
"# Code Coverage Report

| Package |    Coverage   |
| ------- | ------------- |
| main.ts | ğŸ”´ 0% |
| reports/clover/index.ts | ğŸŸ¢ 100% |
| reports/clover/parser/index.ts | ğŸŸ¢ 93.93% |
| reports/clover/types/index.ts | ğŸŸ¢ 100% |
| reports/cobertura/index.ts | ğŸŸ¢ 100% |
| reports/cobertura/parser/index.ts | ğŸ”´ 20% |
| reports/cobertura/types/index.ts | ğŸŸ¢ 100% |
| utils.ts | ğŸŸ  67.36% |
| **Overall Coverage** | **ğŸŸ¢ 49.83%** |

_Minimum allowed coverage is_ \`0%\`_, this run produced_ \`49.83%\`
"
`;

exports[`Generate Diffed Clover Markdown 1`] = `
[
  "::debug::headCoverage: 50.51\\n",
  "::debug::baseCoverage: 50.51\\n",
  "::debug::overallDifferencePercentage: 0\\n",
  "::debug::negativeDifferenceThreshold: 0\\n",
  "Writing results to code-coverage-results.md\\n",
  "\\n",
  "::set-output name=file::code-coverage-results.md\\n",
  "\\n",
  "::set-output name=coverage::50.51\\n",
  "Writing job summary\\n",
]
`;

exports[`Generate Diffed Clover Markdown 2`] = `
"# Code Coverage Report

| Package | Base Coverage | New Coverage | Difference |
| ------- | ------------- | ------------ | ---------- |
| main.ts | ğŸ”´ 0% | ğŸ”´ 0% | âšª 0% |
| reports/clover/index.ts | ğŸŸ¢ 100% | ğŸŸ¢ 100% | âšª 0% |
| reports/clover/parser/index.ts | ğŸŸ¢ 84.31% | ğŸŸ¢ 84.31% | âšª 0% |
| reports/clover/types/index.ts | ğŸŸ¢ 100% | ğŸŸ¢ 100% | âšª 0% |
| reports/cobertura/index.ts | ğŸŸ  66.67% | ğŸŸ  66.67% | âšª 0% |
| reports/cobertura/parser/index.ts | ğŸ”´ 13.33% | ğŸ”´ 13.33% | âšª 0% |
| reports/cobertura/types/index.ts | ğŸŸ¢ 100% | ğŸŸ¢ 100% | âšª 0% |
| utils.ts | ğŸŸ  70.94% | ğŸŸ  70.94% | âšª 0% |
| **Overall Coverage** | **ğŸŸ¢ 50.51%** | **ğŸŸ¢ 50.51%** | **âšª 0%** |

_Minimum allowed coverage is_ \`0%\`_, this run produced_ \`50.51%\`
"
`;

exports[`Generate Diffed Cobertura Markdown 1`] = `
[
  "::debug::headCoverage: 49.83\\n",
  "::debug::baseCoverage: 49.83\\n",
  "::debug::overallDifferencePercentage: 0\\n",
  "::debug::negativeDifferenceThreshold: 0\\n",
  "Writing results to code-coverage-results.md\\n",
  "\\n",
  "::set-output name=file::code-coverage-results.md\\n",
  "\\n",
  "::set-output name=coverage::49.83\\n",
  "Writing job summary\\n",
]
`;

exports[`Generate Diffed Cobertura Markdown 2`] = `
"# Code Coverage Report

| Package | Base Coverage | New Coverage | Difference |
| ------- | ------------- | ------------ | ---------- |
| main.ts | ğŸ”´ 0% | ğŸ”´ 0% | âšª 0% |
| reports/clover/index.ts | ğŸŸ¢ 100% | ğŸŸ¢ 100% | âšª 0% |
| reports/clover/parser/index.ts | ğŸŸ¢ 93.93% | ğŸŸ¢ 93.93% | âšª 0% |
| reports/clover/types/index.ts | ğŸŸ¢ 100% | ğŸŸ¢ 100% | âšª 0% |
| reports/cobertura/index.ts | ğŸŸ¢ 100% | ğŸŸ¢ 100% | âšª 0% |
| reports/cobertura/parser/index.ts | ğŸ”´ 20% | ğŸ”´ 20% | âšª 0% |
| reports/cobertura/types/index.ts | ğŸŸ¢ 100% | ğŸŸ¢ 100% | âšª 0% |
| utils.ts | ğŸŸ  67.36% | ğŸŸ  67.36% | âšª 0% |
| **Overall Coverage** | **ğŸŸ¢ 49.83%** | **ğŸŸ¢ 49.83%** | **âšª 0%** |

_Minimum allowed coverage is_ \`0%\`_, this run produced_ \`49.83%\`
"
`;

exports[`Only list changed files 1`] = `
[
  "::debug::headCoverage: 49\\n",
  "::debug::baseCoverage: 50.51\\n",
  "::debug::overallDifferencePercentage: -1.51\\n",
  "::debug::negativeDifferenceThreshold: 0\\n",
  "Writing results to code-coverage-results.md\\n",
  "\\n",
  "::set-output name=file::code-coverage-results.md\\n",
  "\\n",
  "::set-output name=coverage::49\\n",
  "Writing job summary\\n",
]
`;

exports[`Only list changed files 2`] = `
"# Code Coverage Report: Only Changed Files listed

| Package | Base Coverage | New Coverage | Difference |
| ------- | ------------- | ------------ | ---------- |
| utils.ts | ğŸŸ  70.94% | ğŸŸ  69% | ğŸ”´ -1.94% |
| **Overall Coverage** | **ğŸŸ¢ 50.51%** | **ğŸŸ¢ 49%** | **ğŸ”´ -1.51%** |

_Minimum allowed coverage is_ \`0%\`_, this run produced_ \`49%\`
"
`;
